src/librustc_ast_passes/ast_validation.rs

```diff
@@ -816,44 +816,47 @@ impl<'a> Visitor<'a> for AstValidator<'a> {
                 constness,
                 generics: _,
                 of_trait: None,
-                self_ty: _,
+                ref self_ty,
                 items: _,
             } => {
+                let error = |annotation_span, annotation, note, code| {
```

Comment:
Packing everything into a closure doesn't feel good to me, especially since you have to use two conditionals. I would prefer just extracting the common parts into a method (without the conditionals), returning a builder, and then making necessary special cased modifications on that builder (+ emitting).

Reply:
Any that point you're only avoiding repetition of one line or two.

Reply:
I'm less interested in counting +/- LOCs than how much semantic repetition there is, so even two lines is fine?